tools/chapel-py/method-tables.h

```diff
@@ -0,0 +1,461 @@
+/*
+ * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// This file defines nearly all the methods exposed to python from the frontend.
+//
+// To use these definitions, define the macro and then `#include` this file.
+// See `chapel.cpp` for examples. If the macros below are not defined, they are
+// defined to be empty. Only the ones that are used at the time of include need
+// to be defined.
+//
+
+//
+// Defines the beginning of an ast node
+//
+#ifndef METHOD_TABLE_BEGIN
+#define METHOD_TABLE_BEGIN(TAG)
+#endif
+
+//
+// Defines the end of an ast node
+//
+#ifndef METHOD_TABLE_END
+#define METHOD_TABLE_END(TAG)
+#endif
+
+//
+// Defines a simple getter that takes no arguments
+//
+// BODY is an inline definition of the getter. Available variables:
+// - contextObject: the current frontend context
+// - node: the ast node
+// TYPESTR is the return type. Possible values:
+// - "O" = python object
+// - "s" = string
+// - "b" = bool
+//
+#ifndef PLAIN_GETTER
+#define PLAIN_GETTER(NODE, NAME, DOCSTR, TYPESTR, BODY)
+#endif
+
+//
+// Declares a simple getter that takes no arguments. The definition is defined
+// elsewhere in a function named `NODEObject_NAME`. See the `ACTUAL_ITERATOR`
+// macro for an example
+//
+#ifndef METHOD_PROTOTYPE
+#define METHOD_PROTOTYPE(NODE, NAME, DOCSTR)
+#endif
+
+
+//
+// The order here should be kept in sync with the order in the uast list.
+// Not all nodes need methods exposed to python, so not all uast nodes are listed here.
+// See list in frontend/include/chpl/uast/uast-classes-list.h.
+//
+// Inside each method table, methods should be listed in alphabetical order
+//
+
+METHOD_TABLE_BEGIN(AnonFormal)
```

Comment:
You know, if you called `METHOD_TABLE_BEGIN` something like `CLASS_BEGIN`, it would look pretty much like a DSL for declaring classes. Strictly speaking it wouldn't be accurate (classes are auto-defined even if they have no special methods), but it might look nice.